### ISSUE TITLE
ajax_alert helper methods deprecated

### ISSUE THREAD

**Issue by njlxyaoxinwei** [YaleSTC/shifts#383]
`ajax_alert` and `persistent_ajax_alert` are deprecated methods with the page object.
They were used to display an error message at the top right corner of the page.
A code base search shows that they are still used for the following situtations:
1. In application_controller, they are used in the ajax response of a bunch of (8!) require-admin (or require-owner) validation checks. It is not yet clear if those actions would be called with ajax in Shifts application
2. In all the displays of tooltip form validation errors, including:
   - `RepeatingEventsController`'s `create` and `update` actions for creating and updating recurring shifts and timeslots. However, right now all repeating_event forms --creating recurring shifts and timeslots with tooltip.js, and editing recurring shifts with tooltip (there is no way to edit repeating timeslots, only individual timeslots) --are changed to non-ajax so this problem may not manifest itself.
   - `ShiftsController`'s `show` (???), `create`, `update` and `destroy` actions. Error message for creating/editing/destroying shifts with tooltip.js depends on this deprecated `ajax_alert`. I am not sure why `show` actions has an ajax response. I guess maybe when someone clicks a shift that does not exists anymore (then it shouldn't show up in the first place) on the schedule box, instead of showing the normal shift report in the tooltip, an ajax_alert should occur.
   - `TimeSlotsController`'s `create` and `update` action. Error message for creating/editing timeslots with tooltip.js depend on this. 
3. `StatsController`'s `show` (???) action. I'm not sure where it is needed. I am not even sure if this entire action is needed at all. On the Shifts 2 DEMO site, a stat show page is the same as the shift show page, except with fewer links. If the corresponding shift is not found, for a non-ajax request Shifts 2 DEMO just responds with jQuery code. Example links on Shifts 2 DEMO: [This is a link to a stat show page](http://weke.its.yale.edu/stc_training/stats/100), [this is a link to the corresponding shift show page](http://weke.its.yale.edu/stc_training/shifts/100), and [this is a link to the stat show page of a non-existent shift](http://weke.its.yale.edu/stc_training/stats/10)

In short, the tooltip.js form validation errors won't display correctly until we get rid of all these `ajax_alert`s and find an alternative. I think we can just use the flash notice area of Bootstrap, or alternatively, we can rewrite this as a global JavaScript function and call it in js.erb view files.

PS: Since tooltip.js was the default way to interact with models like Shift and TimeSlot in Shifts 2, the new/edit pages for these models are very incomplete in functionality and, furthermore, do not really have form validation error displays.

**Comment by njlxyaoxinwei** [comment]
In Shifts v3.0.0-rc1 I dealt with three cases by passing an `@ajax_alert_message` to `update.js.erb` for the error to be handled there. The three cases are:
- `ApplicationController#require_department_admin`
- `TimeSlot#update`
- `Shift#update`

The error handling, since they are all kind of the same, should be refactored into one single js file. I will look more into how to achieve this, probably through a shared application controller method.

**Comment by shippy** [comment]
@njlxyaoxinwei If I'm reading this (and the PR) correctly, are these the things to be done?

* Refactoring the AJAX error handling into a single file
* Writing tests that errors are raised and dismissed

Or does the replacement of the original deprecated functions need to be performed in `RepeatingEventsController#create` and `#update`; `ShiftsController#show`, `#create`,  and `#destroy`; and `StatsController#show`?

**Comment by njlxyaoxinwei** [comment]
@shippy That is NOT correct.

Sorry about the confusion. AJAX alert should have already been refactored into one file: ajax_alert.js.erb in the layout view. The shared application controller method is also implemented. There should also already be tests for the error messages. I think the error handling in repeatingevents is already done too.

What I did not finish on branch 383_ajax_alert is not so much the issue described in #383 but rather making all tooltip interactions ajax so there is no full page reload after pressing buttons like destroy/create etc. Granted these two are different things, but I did not create a new issue nor switch to a new branch because I sort of underestimated the work required for this change and thought I could finish by the end of the semester, and these two things are tightly coupled.
